docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,244 @@
+# 009. LNS sticky affinity over multiple sessions
+
+**Feature**: [#1475](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1475)  
+
+**Authors**:
+
+**Status**: Proposed
+__________
+
+## Problem statement
+
+Consider the scenario:
+
+```mermaid
+flowchart LR;
+  Device-->LBS1-->LNS1--1-->IoTHub;
+  Device-->LBS2-->LNS2--2-->IoTHub;
+```
+
+where LNS 1 and 2 make use of their IoT Edge Hub modules to connect to IoT Hub.
+
+IoT Hub limits active connections that an IoT device can have to one. Imagining that connection 1 is
+already open and a message from LNS2 arrives, IoT Hub will close connection 1 and open connection 2.
+Edge Hub on LNS1, will detect this and assume it's a transient network issue, therefore will try
+proactively to reconnect to IoT Hub. IoT Hub will now drop the connection 2 to re-establish the
+original connection 1.  
+
+This connection "ping-pong" will continue happening, negatively impacting the scalability due to the
+high costs of setting up/disposing the connections. From our load tests we observed that in this
+scenario we were not even able to connect more than 120 devices to two LNSs, while in a single LNS
+topology we could scale up to 900 devices without issues.
+
+## Out of scope
+
+- Deduplication strategies Mark and None: these strategies rely on multiple LNSs sending message.
+Potentially we could consider other work arounds for the IoT Hub limitation of a single connection
+per device but we find it acceptable for the Mark and None strategies to not be as scalable as the
+Drop strategy and will only document this limitation for potential users to be aware of.
+
+- LNS performs operations on behalf of a device/sensor and a concentrator/station. However since a
+concentrator can be connected to at most one LNS, there is no ping-pong happening with operations on
+stations.
+
+## Limitations
+
+- We should avoid introducing additional calls to the Function as this would also hurt scalability.
+  
+## In-scope
+
+- The problem can be manifested whenever we do operations against Iot Hub on behalf of edge devices.
+  These can be:
+  - Twin reads
+  - Twin writes (updates/deletes)
+  - D2C messages
+  - C2D messages
+- Roaming edge devices (that potentially lose connectivity to an LNS) are in scope.
+
+### Problematic connections opened on behalf of edge devices
+
+- Data message flows
+  - Join: 
+  - Data: MessageDispatcher.DispatchLoraDataMessage gets the twin (if LoRaDevice not in cache)
```

Comment:
both join and data get the twins. That's the case anytime we create and initialize a device that is not in the cache.

Reply:
The section is re-written but to your comment I think for joins, we first call `loRaDeviceAPIService.SearchAndLockForJoinAsync` before getting the twin, which means if the DevNonce is already used we don't get the twin.

Reply:
absolutely, however, there are edge cases, where the load of the twin for a stale dev address matching your join deveui could conflict with the load you are doing for the join (resoling to the same devaddr). It's an edge case, but simply by protecting using the devnonce, you are not protected against all races when loading the twins. Something Atif is looking into.